Solved issue #118 "Missing VERSION_INFO when compiling as Windows DLL"

Blonder opened the pull request:
This solves JuliaLang/utf8proc issue #118 "Missing VERSION_INFO when compiling as Windows DLL" using CMakeHelpsers from https://github.com/halex2005/CMakeHelpers. CMakeHelpers must be place in the same directory as / parallel to utf8proc.
For English resource language you need to replace the following things in the CMakeHelpers/VersionResource.rc file:

FILEFLAGSMASK 0x3fL                    with   FILEFLAGSMASK 0x17
LANGUAGE LANG_RUSSIAN, SUBLANG_DEFAULT with   LANGUAGE LANG_ENGLISH, SUBLANG_DEFAULT
BLOCK "041904b0                        with   BLOCK "040704b0"
VALUE "Translation", 0x419, 1200       with   VALUE "Translation", 0x407, 1200

stevengj commented:
I don't like the idea of the build script depending on CMakeHelpers being installed in a particular directory outside the source tree.

CMakeHelpers is MIT-licensed, and https://github.com/halex2005/CMakeHelpers/blob/master/generate_product_version.cmake hasn't changed in 3 years.  Maybe we should just include it in utf8proc?

Blonder commented:
First I had the same idea to include CMakeHelpers in utf8proc but then I choose the parallel solution because this way I can use it for other projects too where the version info is needed and to be completely independent (no "copy/paste" code that needs to be merged when changes occur - especially important if more than one project uses CMakeHelpers). The CMakeLists.txt file checks if CMakeHelpers is installed and uses it to generate and add the version info or no version info is generated / added.

stevengj commented:
Bundling a version that is known to work for utf8proc will be much less fragile.

Blonder commented:
I updated the VERSION_INFO to 2.4.0

stevengj commented on CMakeLists.txt:
```diff
@@ -10,16 +10,27 @@ project (utf8proc C)
-set(SO_MINOR 3)
-set(SO_PATCH 1)
+set(SO_MINOR 4)
+set(SO_PATCH 0)
```
As the comment says, these numbers indicate ABI compatibility and don't follow the API version number exactly.  From 2.3 to 2.4 there was no change in ABI which is why the minor version stayed 3.

stevengj replied:
Though it's probably easier for us to just match the API number so that we don't have this discussion repeatedly.

Blonder commented:
I refactored the version info problem on Windows and removed the definitions from CMakeLists.txt and moved it to utf8proc.h where the UTF8PROC_VERSION_MAJOR, UTF8PROC_VERSION_MINOR and UTF8PROC_VERSION_PATCH is defined.

Blonder commented:
Do you have any objects to this new solution?
